src/Couchbase.Lite.Shared/Serialization/FleeceMutableDictionary.cs

```diff
@@ -210,33 +209,59 @@ private FLDictIterator BeginIteration()
             // I hate this dance...but it's necessary to convince the commpiler to let
             // me use unsafe methods inside of a generator method
             var i = BeginIteration();
-            
+
             do {
                 var got = Get(i);
                 if (got.Key != null) {
                     yield return got;
                 }
             } while (Advance(ref i));
+
+            EndIteration(i);
         }
 
         private void SetInMap(string key, MValue val)
-        {
-            _newKeys.Add(key);
+        { 
+            using (var encoded = val.NativeObject.FLEncode()) {
+                //Convert object into FLValue
+                var flValue = NativeRaw.FLValue_FromData((FLSlice)encoded, FLTrust.Trusted);
+                Native.FLSlot_SetValue(Native.FLMutableDict_Set(_dict, key), flValue);
+            }
             _map[key] = val;
         }
 
         #endregion
 
         #region Overrides
 
+        public override void InitAsCopyOf(MCollection original, bool isMutable)
+        {
+            base.InitAsCopyOf(original, isMutable);
+            var d = original as FleeceMutableDictionary;
+            _dict = d != null ? d._dict : null;
+            _map = d?._map;
+        }
+
+        protected override void InitInSlot(MValue slot, MCollection parent, bool isMutable)
+        {
+            base.InitInSlot(slot, parent, isMutable);
+            var baseDict = Native.FLValue_AsDict(slot.Value);
+            _dict = Native.FLDict_MutableCopy(baseDict, FLCopyFlags.DefaultCopy);
+            _releaseRequired = true;
+        }
+
+        #endregion
+
+        #region IFLEncodable
+
         public override void FLEncode(FLEncoder* enc)
         {
             if (!IsMutated) {
-                if (_dict == null) {
-                    Native.FLEncoder_BeginDict(enc,0U);
+                if ((FLDict*)_dict == null) {
+                    Native.FLEncoder_BeginDict(enc, 0U);
                     Native.FLEncoder_EndDict(enc);
                 } else {
-                    Native.FLEncoder_WriteValue(enc, (FLValue*) _dict);
+                    Native.FLEncoder_WriteValue(enc, (FLValue*)(FLDict*)_dict);
```

Comment:
Is a direct cast to `FLValue *` not working here?  Because that seems like it would be fine.